File: prover/src/lfm/transcript_kats.rs
//! Transcript KATs for the LFM compress-chain Fiat–Shamir transcript, at 6 and
//! 7 rounds.
//!
//! GENERATED — do not hand-edit. The INPUTS come from
//! `thoughts/shared/lfm-real-hash/transcript-spec/transcript_kats.json`, which
//! the oracle produced from a Python reference written **before any Rust
//! existed**. That ordering is the point: these vectors are a specification the
//! implementation is checked against, not a recording of what the
//! implementation happened to do.
//!
//! ⚠ **The results were re-pinned when the socket widened to twelve lanes**, by
//! `leaf-spec/rate4_kat_gen.py` out of the same oracle. All 12 moved and no
//! input did: `block_len` is `v[14]` and cannot be made mode-dependent, so the
//! transcript domain re-blesses alongside the leaf domain that needed the width
//! (COMMIT.md §1.4.4 H9).
//!
//! Framing (transcript spec §1.2 at the COMMIT.md §1.2 width): identical to the
//! Merkle socket in every respect except the tag word, which is `"LFMT"`
//! instead of `"LFMC"`. So h = BLAKE3_IV, m[0..4] = state, m[4..8] = operand,
//! m[8..12] = 0 — the third input cell, which the unread-`IN` pins force to
//! zero — m[12] = "LFMT", m[13..16] = 0, t = 0, block_len = 52, flags = 0x0B,
//! digest = out[0..4]. At 7 rounds a step is still literally
//! `blake3::hash(state ‖ operand ‖ 0^16 ‖ "LFMT")[..16]`.

/// One transcript step: state, operand, and the resulting state at each round
/// count.
pub struct StepVector {
    pub name: &'static str,
    pub state: [u32; 4],
    pub operand: [u32; 4],
    /// Result at 6 rounds (the A6R variant; no library computes it).
    pub result_6: [u32; 4],
    /// Result at 7 rounds — `blake3::hash(state ‖ operand ‖ "LFMT")[..16]`.
    pub result_7: [u32; 4],
}

pub const STEP_VECTORS: [StepVector; 6] = [
    StepVector {
        name: "zero_state_zero_operand",
        state: [0x00000000, 0x00000000, 0x00000000, 0x00000000],
        operand: [0x00000000, 0x00000000, 0x00000000, 0x00000000],
        result_6: [0x58A784C6, 0xCA20122A, 0x574D1385, 0x4C7F61AC],
        result_7: [0xBB5DF0AD, 0xBB660FC6, 0x401C1FAD, 0x651C297C],
    },
    StepVector {
        name: "zero_state_main_root",
        state: [0x00000000, 0x00000000, 0x00000000, 0x00000000],
        operand: [0x01020304, 0x05060708, 0x090A0B0C, 0x0D0E0F10],
        result_6: [0xBFD9E2ED, 0x726EDE27, 0x91805DE1, 0xC11F0DA8],
        result_7: [0x503FDDF4, 0x48633531, 0x8EEA401C, 0x213213C8],
    },
    StepVector {
        name: "ramp_state_ramp_operand",
        state: [0x01020304, 0x05060708, 0x090A0B0C, 0x0D0E0F10],
        operand: [0x11121314, 0x15161718, 0x191A1B1C, 0x1D1E1F20],
        result_6: [0x00A8B31B, 0x0C48A09A, 0x1D06A9A8, 0x6C27BD61],
        result_7: [0x29D95598, 0x69E4FD73, 0x243BFCE9, 0x14598F96],
    },
    StepVector {
        name: "max_state",
        state: [0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF],
        operand: [0xDEADBEEF, 0xCAFEBABE, 0x8BADF00D, 0xFEEDFACE],
        result_6: [0xA710A43E, 0x62E96839, 0xE00D7CA2, 0x1E054FEF],
        result_7: [0xBA98C5EF, 0xAFDC8C3E, 0xFA425A12, 0xF35C1B47],
    },
    StepVector {
        name: "squeeze_operand_0",
        state: [0x01020304, 0x05060708, 0x090A0B0C, 0x0D0E0F10],
        operand: [0x305A5153, 0x00000000, 0x00000000, 0x00000000],
        result_6: [0xE1F1E0DF, 0x9B1491E4, 0x26F46CE4, 0x644BA9F0],
        result_7: [0x2EBCFDA8, 0x2F7C4E72, 0xAE841641, 0x6751FE80],
    },
    StepVector {
        name: "squeeze_operand_255",
        state: [0x11121314, 0x15161718, 0x191A1B1C, 0x1D1E1F20],
        operand: [0x305A5153, 0x000000FF, 0x00000000, 0x00000000],
        result_6: [0xDD47DC57, 0x9AC95714, 0xE774A0DA, 0xD4703C0B],
        result_7: [0x33396F61, 0x832BA04F, 0x2BB788AB, 0xE9FE006B],
    },
];

/// The END-TO-END vector: a `FriToyV0`-preamble-shaped transcript, op by op.
///
/// The operation sequence is fixed and lives in the test that replays it —
/// `absorb(main_root), squeeze, squeeze, absorb(l1_root), squeeze,
/// absorb_felts(t0w), absorb_felts(t1w), 4× squeeze_bits`, ✓ VERIFIED against
/// `programs::fri_toy_program_source`. What is pinned here is the STATE after
/// every recorded op, so a divergence is located at the step it happened rather
/// than at the end.
///
/// The last two absorbs are `absorb_felts`, not `absorb2`: the terminal
/// coefficients are field DATA, so each is leaf-hashed under `"LFML"` and the
/// DIGEST is absorbed. The transcript's step count is the same either way, which
/// is why this vector had to be re-pointed deliberately when the program moved
/// rather than caught by a red test.
pub struct EndToEndVector {
    /// State after each recorded op, in order.
    pub states: [[u32; 4]; 11],
    /// The three ext challenges (lanes 0–2 of a squeezed cell).
    pub alpha: [u32; 3],
    pub zeta0: [u32; 3],
    pub zeta1: [u32; 3],
    /// `QUERY_BITS` index bits per query, low-to-high.
    pub query_bits: [[u8; 4]; 4],
}

/// The transcript's inputs — the four cells the preamble absorbs.
pub const MAIN_ROOT: [u32; 4] = [0x01020304, 0x05060708, 0x090A0B0C, 0x0D0E0F10];

pub const L1_ROOT: [u32; 4] = [0x11121314, 0x15161718, 0x191A1B1C, 0x1D1E1F20];

pub const T0W: [u32; 4] = [0xDEADBEEF, 0xCAFEBABE, 0x8BADF00D, 0xFEEDFACE];

pub const T1W: [u32; 4] = [0x0BADC0DE, 0xD15EA5E5, 0xC0FFEE00, 0xBAAAAAAD];

/// Compressions the whole preamble costs — the oracle's cost claim.
///
/// **13, not 11.** Eleven are TRANSCRIPT steps (5 absorbs, 6 squeezes); the
/// other two are the LEAF rows `absorb_felts` adds, one per data cell. Counting
/// them here is the oracle's convention and it is the one that closes the
/// `FriToyV0` total: 4 queries × 20 + 13 = 93.
pub const FRI_TOY_COMPRESSIONS: usize = 13;

/// The end-to-end vector at 7 rounds (the default build).
pub const FRI_TOY_7: EndToEndVector = EndToEndVector {
    states: [
        [0x503FDDF4, 0x48633531, 0x8EEA401C, 0x213213C8],
        [0x02FEA9A6, 0xE6BF885C, 0xF174E65F, 0x4EC9AB10],
        [0xF91F56DE, 0x62F37956, 0xE67D5421, 0xD82727D0],
        [0x7A31B840, 0xAD7F2625, 0xE27D1C56, 0xCDB0E9A7],
        [0xA6790B7B, 0x00695D49, 0xA663DC33, 0x2E849F0C],
        [0xFC40465E, 0x0092B147, 0x2FA48645, 0x9755608B],
        [0x6BD6F0B0, 0x634CF1C6, 0x3CBD9D2D, 0x349F278B],
        [0xD86E1D3F, 0xDD1CFBC3, 0x1C8E8F14, 0x22D35494],
        [0x5B449138, 0x3435B7D5, 0x7CFE4C06, 0x1C022FCF],
        [0xE8D9848C, 0x0429B6F7, 0xDD5CBA1A, 0xBD465F16],
        [0xD1E3472D, 0xD945386A, 0xC746A9B3, 0x8FD73C31],
    ],
    alpha: [0x503FDDF4, 0x48633531, 0x8EEA401C],
    zeta0: [0x02FEA9A6, 0xE6BF885C, 0xF174E65F],
    zeta1: [0x7A31B840, 0xAD7F2625, 0xE27D1C56],
    query_bits: [[0, 0, 0, 0], [1, 1, 1, 1], [0, 0, 0, 1], [0, 0, 1, 1]],
};

/// The end-to-end vector at 6 rounds (`--features blake3-6round`).
pub const FRI_TOY_6: EndToEndVector = EndToEndVector {
    states: [
        [0xBFD9E2ED, 0x726EDE27, 0x91805DE1, 0xC11F0DA8],
        [0x9DDE5E52, 0xD7018142, 0x978528FF, 0xA01782B7],
        [0x189BAD31, 0x364E3C30, 0x6B4D5516, 0x78D7FE7B],
        [0x8E3E80BB, 0xFCA1AF96, 0xA59E0F41, 0x18C9AB19],
        [0x4DF9BD75, 0x2E131E7D, 0x2DEB348E, 0x62BC30F3],
        [0x202DB868, 0x8FF72AC4, 0x452D536A, 0x78DECD7B],
        [0xA3D1E95E, 0x8C68ECF6, 0x91D7DF5E, 0x28D8CBB2],
        [0xFA4222A2, 0x18DA7862, 0x3BBDA144, 0xAB453013],
        [0x525CE059, 0x0E9B9EAB, 0xFB57633B, 0xE43490A7],
        [0x261309AA, 0x84C675B7, 0xD5BBF0EB, 0x4AF40850],
        [0x4D62F602, 0x2CC5D660, 0x99C44AF7, 0xEB502D8A],
    ],
    alpha: [0xBFD9E2ED, 0x726EDE27, 0x91805DE1],
    zeta0: [0x9DDE5E52, 0xD7018142, 0x978528FF],
    zeta1: [0x8E3E80BB, 0xFCA1AF96, 0xA59E0F41],
    query_bits: [[0, 1, 1, 1], [0, 1, 0, 0], [1, 0, 0, 1], [0, 1, 0, 1]],
};

File: thoughts/shared/lfm-real-hash/transcript-spec/transcript_kats.py
"""
Transcript KATs: per-op vectors, an END-TO-END FriToyV0-shaped transcript, and
the checks that make them meaningful.

What this establishes, all ✓ EXECUTED:
  K1  every transcript step equals `BLAKE3(state ‖ operand ‖ "LFMT")[..16]` at
      7 rounds — computed by two separate routes (word level and byte level)
      and asserted equal. This is the crate-KAT identity the implementer must
      re-assert with a one-line `blake3::hash` call.
  K2  a full FriToyV0-preamble-shaped transcript, op by op, with the state
      after every step — so the implementer has an end-to-end vector, not only
      per-op ones.
  K3  DOMAIN SEPARATION IS REAL: a transcript step and a Merkle parent over the
      same two cells produce different digests (the tag is load-bearing).
  K4  the squeeze counter is load-bearing: dropping it makes consecutive
      squeezes iterate one fixed map, and the vectors change.
  K5  ordering is load-bearing: swapping two absorbs changes the transcript.
  K6  compression accounting matches the spec's cost claim (11 for FriToyV0).

Run: python3 transcript_kats.py [--write]
"""

from __future__ import annotations

import json
import os
import sys

import transcript_ref as tr

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                "..", "gate-oracle"))
import socket_ref as sk              # noqa: E402

HERE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(HERE, "transcript_kats.json")

# FriToyV0's shape, ✓ VERIFIED against fixture.rs:26-40.
NUM_QUERIES = 4
QUERY_BITS = 4

# Fixed, written-out inputs — nothing depends on an RNG.
MAIN_ROOT = [0x01020304, 0x05060708, 0x090A0B0C, 0x0D0E0F10]
L1_ROOT = [0x11121314, 0x15161718, 0x191A1B1C, 0x1D1E1F20]
T0W = [0xDEADBEEF, 0xCAFEBABE, 0x8BADF00D, 0xFEEDFACE]
T1W = [0x0BADC0DE, 0xD15EA5E5, 0xC0FFEE00, 0xBAAAAAAD]


def hexlanes(c):
    return "".join(f"{x:08x}" for x in c)


def k1_step_identity(rounds: int) -> tuple[bool, str, list]:
    """Every step, both routes. At 7 rounds the byte route IS `blake3::hash`."""
    cases = [
        ("zero_state_zero_operand", [0, 0, 0, 0], [0, 0, 0, 0]),
        ("zero_state_main_root", [0, 0, 0, 0], MAIN_ROOT),
        ("ramp_state_ramp_operand", MAIN_ROOT, L1_ROOT),
        ("max_state", [0xFFFFFFFF] * 4, T0W),
        ("squeeze_operand_0", MAIN_ROOT, tr.squeeze_operand(0)),
        ("squeeze_operand_255", L1_ROOT, tr.squeeze_operand(255)),
    ]
    out = []
    for name, st, op in cases:
        w = tr.compress_t(st, op, rounds)
        b = tr.compress_t_bytelevel(st, op, rounds)
        if w != b:
            return False, f"K1 route mismatch on {name}@{rounds}", []
        out.append({"name": name, "state": st, "operand": op,
                    "result": w, "result_hex": hexlanes(w)})
    return True, f"K1 PASS: {len(cases)} steps, word route == byte route", out


def k2_end_to_end(rounds: int) -> tuple[dict, list]:
    """The FriToyV0 preamble + query loop, op by op.

    ✓ VERIFIED sequence, programs.rs:549-567:
      absorb(main_root), squeeze_ext, squeeze_ext, absorb(l1_root),
      squeeze_ext, absorb_felts(t0w), absorb_felts(t1w), then
      NUM_QUERIES x squeeze_bits.
    """
    t = tr.Transcript(rounds=rounds)
    steps = []

    def rec(op, value=None):
        steps.append({"op": op,
                      "state_after": list(t.state),
                      "state_after_hex": hexlanes(t.state),
                      **({"output": value, "output_hex": hexlanes(value)}
                         if value is not None and len(value) == 4 else {}),
                      **({"output_lanes": value} if value is not None
                         and len(value) != 4 else {})})

    t.absorb(MAIN_ROOT);            rec("absorb(main_root)")
    alpha = t.squeeze_ext();        rec("squeeze_ext -> alpha", alpha)
    zeta0 = t.squeeze_ext();        rec("squeeze_ext -> zeta0", zeta0)
    t.absorb(L1_ROOT);              rec("absorb(l1_root)")
    zeta1 = t.squeeze_ext();        rec("squeeze_ext -> zeta1", zeta1)
    # ✓ VERIFIED programs.rs: the preamble now calls absorb_felts TWICE, not
    # absorb2 — t0/t1 are terminal-polynomial coefficients, i.e. ARBITRARY field
    # elements, so each goes leaf-then-absorb. Four compresses where the old
    # vector modelled two; this is the 91 -> 93 correction, in the vectors.
    t.absorb_felts(T0W);            rec("absorb_felts(t0w)")
    t.absorb_felts(T1W);            rec("absorb_felts(t1w)")
    query_bits = []
    for q in range(NUM_QUERIES):
        bits = t.squeeze_bits(QUERY_BITS)
        query_bits.append(bits)
        rec(f"squeeze_bits(q={q})", bits)

    return {
        "rounds": rounds,
        "inputs": {"main_root": MAIN_ROOT, "l1_root": L1_ROOT,
                   "t0w": T0W, "t1w": T1W},
        "shape": {"num_queries": NUM_QUERIES, "query_bits": QUERY_BITS},
        "steps": steps,
        "challenges": {"alpha": alpha, "zeta0": zeta0, "zeta1": zeta1,
                       "query_bits": query_bits},
        "compressions": t.compressions,
        "final_state": list(t.state),
    }, steps


def k3_domain_separation(rounds: int) -> tuple[bool, str]:
    """A transcript step must NOT equal a Merkle parent over the same cells."""
    a, b = MAIN_ROOT, L1_ROOT
    step = tr.compress_t(a, b, rounds)
    parent = sk.socket_digest_wordlevel(a, b, sk.Framing(rounds=rounds))
    if step == parent:
        return False, ("K3 FAIL: transcript step == Merkle parent — the tag is "
                       "NOT separating the domains")
    return True, ("K3 PASS: transcript step != Merkle parent on the same two "
                  "cells (the LFMT/LFMC tag is load-bearing)")


def k4_counter_is_load_bearing(rounds: int) -> tuple[bool, str]:
    """Without SQ(i)'s counter every squeeze advance uses ONE fixed operand, so
    a run of squeezes iterates one fixed map. The vectors must notice."""
    t1 = tr.Transcript(rounds=rounds)
    t1.absorb(MAIN_ROOT)
    with_counter = [t1.squeeze() for _ in range(4)]

    t2 = tr.Transcript(rounds=rounds)
    t2.absorb(MAIN_ROOT)
    fixed = tr.squeeze_operand(0)
    without = []
    for _ in range(4):
        without.append(list(t2.state))
        t2.state = tr.compress_t(t2.state, fixed, rounds)

    if with_counter == without:
        return False, "K4 FAIL: the squeeze counter changes nothing"
    first_diff = next(i for i, (x, y) in enumerate(zip(with_counter, without))
                      if x != y)
    return True, (f"K4 PASS: counter-free squeezes diverge from the spec at "
                  f"squeeze #{first_diff} (they iterate one fixed map)")


def k5_order_is_load_bearing(rounds: int) -> tuple[bool, str]:
    a = tr.Transcript(rounds=rounds); a.absorb(MAIN_ROOT); a.absorb(L1_ROOT)
    b = tr.Transcript(rounds=rounds); b.absorb(L1_ROOT); b.absorb(MAIN_ROOT)
    if a.state == b.state:
        return False, "K5 FAIL: absorb order does not affect the transcript"
    return True, "K5 PASS: swapping two absorbs changes the state"


def main() -> int:
    print("=" * 74)
    print("TRANSCRIPT KATs — compress-chain (option B1)")
    print("=" * 74)
    ok = True
    doc = {"construction": "LFM compress-chain transcript (option B1)",
           "tag_ascii": tr.TAG_LFMT_ASCII.decode(),
           "tag_word": tr.TAG_LFMT,
           "squeeze_mark": tr.SQUEEZE_MARK,
           "state_cells": 1,
           "state_bits": 128,
           "initial_state": tr.ZERO_CELL,
           "framing": "identical to the Merkle socket except m[8] = TAG_LFMT",
           "rounds": {}}

    for rounds in (7, 6):
        good, msg, steps = k1_step_identity(rounds)
        ok &= good
        print(f"  [{'PASS' if good else 'FAIL'}] {msg}")
        e2e, _ = k2_end_to_end(rounds)
        doc["rounds"][str(rounds)] = {"step_vectors": steps, "fri_toy_v0": e2e}
        print(f"  [PASS] K2: FriToyV0-shaped transcript, {len(e2e['steps'])} "
              f"ops, {e2e['compressions']} compressions @{rounds}r")

    for fn in (k3_domain_separation, k4_counter_is_load_bearing,
               k5_order_is_load_bearing):
        good, msg = fn(7)
        ok &= good
        print(f"  [{'PASS' if good else 'FAIL'}] {msg}")

    # K6 — the cost claim in the spec must match what the reference performed.
    n = doc["rounds"]["7"]["fri_toy_v0"]["compressions"]
    good = (n == 13)
    ok &= good
    print(f"  [{'PASS' if good else 'FAIL'}] K6: FriToyV0 transcript costs "
          f"{n} compressions (13: 11 + the 2 leaf rows)")

    if "--write" in sys.argv:
        with open(OUT, "w") as f:
            json.dump(doc, f, indent=1)
        print(f"\n  wrote {OUT}")

    print("-" * 74)
    print(f"TRANSCRIPT KATs: {'PASS' if ok else 'FAIL'}")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())
